src/httpClient/default.ts

```diff
@@ -0,0 +1,220 @@
+}
+
+async function consumeResponseText(response: Response): Promise<TpufResponseWithMetadata> {
+  const body_text = await response.text();
+  const body_read_end = performance.now();
+  return { body_text, body_read_end, decompress_end: body_read_end };
+}
+
+export default class DefaultHTTPClient implements HTTPClient {
+  private baseUrl: string;
+  private origin: URL;
+  private apiKey: string;
+  private connectTimeout: number;
+  readonly userAgent = `tpuf-typescript/${version}`;
+  private compression: boolean;
+
+  constructor(
+    baseUrl: string,
+    apiKey: string,
+    connectTimeout: number,
+    idleTimeout: number, // not used currently, just to match NodeHTTPClient's signature
+    warmConnections: number,
+    compression: boolean,
+  ) {
+    this.baseUrl = baseUrl;
+    this.origin = new URL(baseUrl);
+    this.origin.pathname = "";
+    this.apiKey = apiKey;
+    this.connectTimeout = connectTimeout;
+    this.compression = compression;
+
+    for (let i = 0; i < warmConnections; i++) {
+      // send a small request to put some connections in the pool
+      void fetch(this.baseUrl, {
+        method: "HEAD",
+        headers: { "User-Agent": this.userAgent },
+      });
+    }
+  }
+
+  async doRequest<T>({
+    method,
+    path,
+    query,
+    body,
+    compress,
+    retryable,
+  }: RequestParams): RequestResponse<T> {
+    const url = new URL(`${this.baseUrl}${path}`);
+    if (query) {
+      Object.keys(query).forEach((key) => {
+        const value = query[key];
+        if (value) {
+          url.searchParams.append(key, value);
+        }
+      });
+    }
+    path = url.pathname;
+    if (query) {
+      path += "?" + url.search;
+    }
+
+    const headers: Record<string, string> = {
+      Authorization: `Bearer ${this.apiKey}`,
+      "User-Agent": this.userAgent,
+    };
+
+    if (this.compression) {
+      headers["Accept-Encoding"] = "gzip";
+    }
+
+    if (body) {
+      headers["Content-Type"] = "application/json";
+    }
+
+    let requestCompressionDuration;
+    let requestBody: Uint8Array | string | null = null;
+    if (body && compress && this.compression) {
+      headers["Content-Encoding"] = "gzip";
+      const beforeRequestCompression = performance.now();
+      requestBody = gzip(JSON.stringify(body));
+      requestCompressionDuration = performance.now() - beforeRequestCompression;
+    } else if (body) {
+      requestBody = JSON.stringify(body);
+    }
+
+    const maxAttempts = retryable ? 3 : 1;
+    let response!: Response;
+    let error: TurbopufferError | null = null;
+    let request_start!: number;
+    let response_start!: number;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      error = null;
+      request_start = performance.now();
+      try {
+        response = await fetch(new URL(path, this.origin), {
+          method,
+          headers,
+          body: requestBody,
+          signal: AbortSignal.timeout(this.connectTimeout),
```

Comment:
I think this timeout will be used for the entire request up until the headers are returned. The connect timeout should only apply for the tcp handshake. Likely this can't be configured for fetch()

oddly we don't have a read_timeout configuration, we're just relying on the default read timeout at the moment

Reply:
[for posterity]
we decided to look at adding this in a separate PR